aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+called, it wouldn't route back through the autocasting logic at all.
+
+Option 2:  Patch the Python-exposed surface of calls, to make 100% sure autograd history
+recording can't sneak in ahead of autocast.  This mirrors Apex most closely.
+
+I think Option 2 is the right answer for all ops, not just convolutions.  Option 2 is what I implement here.
+*****************************************************************************************************************/
+
+auto register_fallthrough = c10::Dispatcher::singleton()
+  .registerBackendFallbackKernel(DispatchKey::AutocastTensorId,
+                                 KernelFunction::makeFallthrough());
+
+// TODO:  Codegen the stuff below?  Ed said
+// > you are going to have to write the function definition at some point, I wouldn't try to get clever about it
+// Therefore, for the moment, this is all copy pasted in from VariableTypeEverything.cpp with appropriate substitutions.
+
+// Macros to reduce boilerplate somewhat.
+// Common case where registration signature matches redispatch signature
+// (that's why SIGNATURE is repeated in the WrapFunction instantiation)
+#define KERNEL(FUNC, REGISTER_SCHEMA, SIGNATURE, POLICY) \
+  .op(torch::RegisterOperators::options() \
+    .schema(REGISTER_SCHEMA) \
+    .kernel<SIGNATURE>(DispatchKey::AutocastTensorId, \
+    &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, FUNC>::type::call) \
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
+
+#define KERNEL_UNBOXED_ONLY(FUNC, REGISTER_SCHEMA, SIGNATURE, POLICY) \
+  .op(torch::RegisterOperators::options() \
+    .schema(REGISTER_SCHEMA) \
+    .impl_unboxedOnlyKernel<SIGNATURE, \
+    &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, FUNC>::type::call \
+    >(DispatchKey::AutocastTensorId) \
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
+
+// Less-common but still useful case: redispatching to a function with a new signature (e.g. appending a dtype)
+#define KERNEL_UNBOXED_ONLY_DIFFERENT_REDISPATCH_SIGNATURE(REDISPATCH_FUNC, REGISTER_SCHEMA, REGISTER_SIGNATURE, REDISPATCH_SIGNATURE, POLICY) \
+  .op(torch::RegisterOperators::options() \
+    .schema(REGISTER_SCHEMA) \
+    .impl_unboxedOnlyKernel<REGISTER_SIGNATURE, \
+    &WrapFunction<CastPolicy::POLICY, REGISTER_SIGNATURE, REDISPATCH_SIGNATURE, REDISPATCH_FUNC>::type::call \
+    >(DispatchKey::AutocastTensorId) \
+    .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
+
+/*****************************************
+Explicit registration for out-of-place ops
+*****************************************/
+auto register_out_of_place = torch::RegisterOperators()
+  // fp16
+  // TODO (to consider): should the slow_conv*d from python_nn_functions_dispatch.h go here?
+  // They use unfolding + admm (which may use cublas) so it's possible they benefit from Tensor Cores,
+  // although if that code path is being used, it's probably unsalvageably slow regardless.
```

Comment:
No users should be directly calling those functions anyway.

Reply:
(They are incorrectly bound without a leading underscore.)

Reply:
just to be sure, is that a "no, you should not register anything for slow_conv*d?"

Reply:
the more comments I can delete, the better chance I have to maintain this PR at exactly 1000 lines.

Reply:
Deleted mention of slow_conv*d in my branch.